### PR TITLE
Service introspection

### PR DESCRIPTION
**Pull request by ihasdapie**
This PR is part of a prototype implementation for https://github.com/ros-infrastructure/rep/pull/360 https://github.com/ros2/ros2/issues/1285. It implements the new methods defined in `rosidl_runtime_cpp` in https://github.com/ros2/rosidl/pull/700 allowing for rcl to populate service_event message payloads.

### CODE REVIEW

File: rosidl_typesupport_cpp/resource/srv__type_support.cpp.em [hunk 1]
```diff
@@ -109,12 +117,86 @@ static const type_support_map_t _@(service.namespaced_type.name)_service_typesup
...
+
 static const rosidl_service_type_support_t @(service.namespaced_type.name)_service_type_support_handle = {
-  ::rosidl_typesupport_cpp::typesupport_identifier,
-  reinterpret_cast<const type_support_map_t *>(&_@(service.namespaced_type.name)_service_typesupport_map),
-  ::rosidl_typesupport_cpp::get_service_typesupport_handle_function,
+  .typesupport_identifier = ::rosidl_typesupport_cpp::typesupport_identifier,
+  .data = reinterpret_cast<const type_support_map_t *>(&_@(service.namespaced_type.name)_service_typesupport_map),
+  .func = ::rosidl_typesupport_cpp::get_service_typesupport_handle_function,
+  .introspection_message_create_handle  = rosidl_@('_'.join([package_name, *interface_path.parents[0].parts, service.namespaced_type.name]))_introspection_message_create,
+  .introspection_message_destroy_handle = rosidl_@('_'.join([package_name, *interface_path.parents[0].parts, service.namespaced_type.name]))_introspection_message_destroy,
```
Comment (jacobperron) [inline rosidl_typesupport_cpp/resource/srv__type_support.cpp.em:189]: What do you think about using term "event message" instead of "introspection message"? It seems a little more precise (and concise). This would mean renaming these handles (and generated functions accordingly) e.g.

```
event_message_create_handle
event_message_destroy_handle
```

File: rosidl_typesupport_cpp/resource/srv__type_support.cpp.em [hunk 2]
```diff
@@ -109,12 +117,86 @@ static const type_support_map_t _@(service.namespaced_type.name)_service_typesup
...
+
+bool
+rosidl_@('_'.join([package_name, *interface_path.parents[0].parts, service.namespaced_type.name]))_introspection_message_destroy(
+void* event_msg,
+rcutils_allocator_t * allocator
+)
+{
+  auto * event_msg_ = static_cast<@event_type *>(event_msg);
+  // event_msg_->@('::'.join([package_name, *interface_path.parents[0].parts]) + '::~' + service.namespaced_type.name + '_Event')();
+  // std::destroy_at(event_msg_); // Why can't I compile with c++17?
```
Comment (jacobperron) [inline rosidl_typesupport_cpp/resource/srv__type_support.cpp.em:177]: I guess if we want to rely on the c-allocators anyways, then we don't need to bother with `std::destroy_at` and can remove these commented lines.

File: rosidl_typesupport_cpp/CMakeLists.txt [hunk 3]
```diff
@@ -2,9 +2,9 @@ cmake_minimum_required(VERSION 3.5)
 
 project(rosidl_typesupport_cpp)
 
-# Default to C++14
+# Default to C++17
 if(NOT CMAKE_CXX_STANDARD)
-  set(CMAKE_CXX_STANDARD 14)
+  set(CMAKE_CXX_STANDARD 17)
```
Comment (jacobperron) [inline rosidl_typesupport_cpp/CMakeLists.txt:7]: I would revert this change if we don't need it.

**Review by jacobperron — COMMENTED** [review]
LGTM, just some minor comments and a bikeshed question about using "event" vs. "introspection".

File: rosidl_typesupport_cpp/resource/srv__type_support.cpp.em [hunk 4]
```diff
@@ -109,12 +120,80 @@ static const type_support_map_t _@(service.namespaced_type.name)_service_typesup
   &_@(service.namespaced_type.name)_service_typesupport_data.data[0],
 };
 
+
+
+#ifdef __cplusplus
```
Comment (jacobperron) [inline rosidl_typesupport_cpp/resource/srv__type_support.cpp.em:125]: Is the `#ifdef` necessary? This is most certainly a cpp file, so maybe we can omit it.

File: rosidl_typesupport_cpp/resource/srv__type_support.cpp.em [hunk 5]
```diff
@@ -109,12 +120,80 @@ static const type_support_map_t _@(service.namespaced_type.name)_service_typesup
...
+  .event_typesupport = ::rosidl_typesupport_cpp::get_message_type_support_handle<@('::'.join([package_name, *interface_path.parents[0].parts, service.namespaced_type.name]))_Event>(),
 };
 
+#ifdef __cplusplus
+}
+#endif
+
+
+
+
```
Comment (jacobperron) [inline rosidl_typesupport_cpp/resource/srv__type_support.cpp.em:196]:
```suggestion
```

File: rosidl_typesupport_cpp/resource/srv__type_support.cpp.em [hunk 6]
```diff
@@ -152,4 +231,8 @@ get_service_type_support_handle<@('::'.join([package_name] + list(interface_path
 @[end if]@
 }
 
+
+
+
+
```
Comment (jacobperron) [inline rosidl_typesupport_cpp/resource/srv__type_support.cpp.em:237]:
```suggestion
```

File: rosidl_typesupport_cpp/src/type_support_dispatch.hpp [hunk 7]
```diff
@@ -108,6 +108,7 @@ get_typesupport_handle_function(
       return ts;
     }
   }
+  rcutils_reset_error();
```
Comment (jacobperron) [inline rosidl_typesupport_cpp/src/type_support_dispatch.hpp:111]: Why reset here? It doesn't seem like we should.

File: rosidl_typesupport_tests/test/rosidl_typesupport_cpp/test_service_typesupport.cpp [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+    &info, &allocator, static_cast<const void *>(&req), nullptr, true));
+
+  EXPECT_EQ(event->info.sequence_number, info.sequence_number);
+  EXPECT_EQ(event->info.event_type, info.event_type);
+  EXPECT_EQ(event->info.stamp.nanosec, info.stamp_nanosec);
+  EXPECT_EQ(event->info.stamp.sec, info.stamp_sec);
+  for (int i = 0; i < 16; ++i) {
+    EXPECT_EQ(event->info.client_id.uuid[i], info.client_id[i]);
+  }
+  EXPECT_EQ(event->request[0].int16_value, -1);
```
Comment (jacobperron) [inline rosidl_typesupport_tests/test/rosidl_typesupport_cpp/test_service_typesupport.cpp:78]:
```suggestion
  ASSERT_EQ(event->request.size(), 1u);
  EXPECT_EQ(event->request[0].int16_value, -1);
```

File: rosidl_typesupport_c/resource/srv__type_support.cpp.em [hunk 9]
```diff
@@ -105,12 +118,84 @@ static const type_support_map_t _@(service.namespaced_type.name)_service_typesup
...
+      event_msg->request.capacity = 1;
+      event_msg->request.size = 1;
+      event_msg->request.data = static_cast<@request_type *>(allocator->zero_allocate(1, sizeof(@request_type), allocator->state));
+      if (! @(request_type)__copy((@request_type *) request_message, &event_msg->request.data[0])){
+        return NULL;
+      }
+    } else {
+      // set error?
+      return NULL;
+    }
```
[the remark this answers is not in the document]
Comment (ihasdapie) [reply rosidl_typesupport_c/resource/srv__type_support.cpp.em:168]: I can't think of any downstream issues either

**Comment by ihasdapie** [comment]
@jacobperron Pinging to request review & help with changes introduced in https://github.com/ros2/rosidl_typesupport/pull/127/commits/d48c0eec70ea057d536a6cd36afc44777b39e6cc to split typesupport tests into a new package so that we can test the generated typesupport files. I'm stuck on some linker issues while building, so some feedback would be much appreciated.

File: rosidl_typesupport_cpp/resource/srv__type_support.cpp.em (hunk 1 above)
Comment (ihasdapie) [reply rosidl_typesupport_cpp/resource/srv__type_support.cpp.em]: Yeah, I agree that it'd be good to rename everything for consistency. This has been resolved in [9e0b9814bcfe332427360b397cdc712fd3aabaf7 ](https://github.com/ros2/rosidl/pull/700/commits/9e0b9814bcfe332427360b397cdc712fd3aabaf7) and 0e03efc8a6f1aa4df8f7b9d023f7eba98a910a3f

File: rosidl_typesupport_c/resource/srv__type_support.cpp.em [hunk 10]
```diff
@@ -105,12 +118,84 @@ static const type_support_map_t _@(service.namespaced_type.name)_service_typesup
...
+
+
+void * rosidl_typesupport_c_@('__'.join([package_name, *interface_path.parents[0].parts, service.namespaced_type.name]))__service_introspection_message__create (
+    const rosidl_service_introspection_info_t * info,
+    rcutils_allocator_t * allocator,
+    const void * request_message,
+    const void * response_message,
+    bool enable_message_payload)
+{
+  auto * event_msg = static_cast<@event_type *>(allocator->zero_allocate(1, sizeof(@event_type), allocator->state));
```
[the remark this answers is not in the document]
Comment (ihasdapie) [reply rosidl_typesupport_c/resource/srv__type_support.cpp.em]: 83301ebfc7ab97309c0e7df7804af291e47346d6

File: rosidl_typesupport_c/resource/srv__type_support.cpp.em [hunk 11]
```diff
@@ -105,12 +118,84 @@ static const type_support_map_t _@(service.namespaced_type.name)_service_typesup
...
+    event_msg->info.client_id.uuid[i] = info->client_id[i];
+  }
+
+  if (enable_message_payload) {
+    if (NULL == request_message) {
+      event_msg->response.capacity = 1;
+      event_msg->response.size = 1;
+      event_msg->response.data = static_cast<@response_type *>(allocator->zero_allocate(1, sizeof(@response_type), allocator->state));
+      if (! @(response_type)__copy((@response_type *) response_message, &event_msg->response.data[0])) {
+        return NULL;
```
[the remark this answers is not in the document]
Comment (ihasdapie) [reply rosidl_typesupport_c/resource/srv__type_support.cpp.em]: 83301ebfc7ab97309c0e7df7804af291e47346d6

File: rosidl_typesupport_c/resource/srv__type_support.cpp.em [hunk 12]
```diff
@@ -105,12 +118,84 @@ static const type_support_map_t _@(service.namespaced_type.name)_service_typesup
...
+  event_msg->info.stamp.nanosec = info->stamp_nanosec;
+  for (size_t i = 0; i < 16; ++i) {
+    event_msg->info.client_id.uuid[i] = info->client_id[i];
+  }
+
+  if (enable_message_payload) {
+    if (NULL == request_message) {
+      event_msg->response.capacity = 1;
+      event_msg->response.size = 1;
+      event_msg->response.data = static_cast<@response_type *>(allocator->zero_allocate(1, sizeof(@response_type), allocator->state));
```
[the remark this answers is not in the document]
Comment (ihasdapie) [reply rosidl_typesupport_c/resource/srv__type_support.cpp.em]: 83301ebfc7ab97309c0e7df7804af291e47346d6

File: rosidl_typesupport_c/resource/srv__type_support.cpp.em [hunk 13]
```diff
@@ -105,12 +118,84 @@ static const type_support_map_t _@(service.namespaced_type.name)_service_typesup
...
+  void * event_msg,
+  rcutils_allocator_t * allocator
+)
+{ 
+  if (NULL == event_msg) {
+    return false;
+  }
+
+  @(event_type)__fini((@event_type *) event_msg);
+  allocator->deallocate(event_msg, allocator->state);
```
[the remark this answers is not in the document]
Comment (ihasdapie) [reply rosidl_typesupport_c/resource/srv__type_support.cpp.em]: 83301ebfc7ab97309c0e7df7804af291e47346d6

File: rosidl_typesupport_cpp/resource/srv__type_support.cpp.em (hunk 4 above)
Comment (ihasdapie) [reply rosidl_typesupport_cpp/resource/srv__type_support.cpp.em]: 96e23a359ffef893646e99216af56499e852e3dd

File: rosidl_typesupport_cpp/src/type_support_dispatch.hpp (hunk 7 above)
Comment (ihasdapie) [reply rosidl_typesupport_cpp/src/type_support_dispatch.hpp]: I remember putting it there to squash a warning but I don't recall what for. I've removed it in 05b9f45c3889b8a87831a3cd758bc9d1303d0106

File: rosidl_typesupport_tests/test/rosidl_typesupport_cpp/test_service_typesupport.cpp (hunk 8 above)
Comment (ihasdapie) [reply rosidl_typesupport_tests/test/rosidl_typesupport_cpp/test_service_typesupport.cpp:80]: 94027049bf9bba649c20d57fee2deee68c37e08c

File: rosidl_typesupport_c/resource/srv__type_support.cpp.em (hunk 9 above)
[the remark this answers is not in the document]
Comment (ihasdapie) [reply rosidl_typesupport_c/resource/srv__type_support.cpp.em]: Resolved this & comment above in 83301ebfc7ab97309c0e7df7804af291e47346d6 (and the zero_allocate review comments...) 

This removes the check and associated error for both request and response members being null: https://github.com/ros2/rosidl_typesupport/blob/d48c0eec70ea057d536a6cd36afc44777b39e6cc/rosidl_typesupport_cpp/resource/srv__type_support.cpp.em#L162

**Comment by jacobperron** [comment]
rebased on rolling

File: rosidl_typesupport_tests/test/rosidl_typesupport_cpp/test_service_typesupport.cpp [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+#include "builtin_interfaces/msg/time.h"
+#include "unique_identifier_msgs/msg/uuid.h"
+#include "service_msgs/msg/service_event_info.h"
+#include "rosidl_runtime_c/service_type_support_struct.h"
+
+TEST(test_service_typesupport, basic_types_cpp_to_cpp)
+{
+  rcutils_allocator_t allocator = rcutils_get_default_allocator();
+  const rosidl_service_type_support_t * srv_ts = 
+    rosidl_runtime_cpp::get_service_type_support_handle<rosidl_typesupport_tests::srv::BasicTypes>();
```
Comment (jacobperron) [inline rosidl_typesupport_tests/test/rosidl_typesupport_cpp/test_service_typesupport.cpp:35]: I guess this should use a different namespace:
```suggestion
    rosidl_typesupport_cpp::get_service_type_support_handle<rosidl_typesupport_tests::srv::BasicTypes>();
```
I don't know _why_, but it seems to fix the linker issue.

Comment (jacobperron) [reply rosidl_typesupport_tests/test/rosidl_typesupport_cpp/test_service_typesupport.cpp]: Fixed in [62b9a1d](https://github.com/ros2/rosidl_typesupport/pull/127/commits/62b9a1d61d53afea91b5eb4b11d35baef19ae56f)

File: rosidl_typesupport_c/resource/srv__type_support.cpp.em (hunk 9 above)
[the remark this answers is not in the document]
Comment (jacobperron) [reply rosidl_typesupport_c/resource/srv__type_support.cpp.em]: There was a bug here where a segfault would occur if given a valid request, but a null response (or vice versa) . Fixed in [62b9a1d](https://github.com/ros2/rosidl_typesupport/pull/127/commits/62b9a1d61d53afea91b5eb4b11d35baef19ae56f)

File: rosidl_typesupport_c/resource/srv__type_support.cpp.em [hunk 15]
```diff
@@ -105,12 +118,92 @@ static const type_support_map_t _@(service.namespaced_type.name)_service_typesup
   &_@(service.namespaced_type.name)_service_typesupport_data.data[0],
 };
 
+
+@#TODO(ihasdapie): Import Postfixes
```
Comment (jacobperron) [inline rosidl_typesupport_c/resource/srv__type_support.cpp.em:122]: I think this was done

File: rosidl_typesupport_c/CMakeLists.txt [hunk 16]
```diff
@@ -9,7 +9,7 @@ endif()
 
 # Default to C++14
 if(NOT CMAKE_CXX_STANDARD)
-  set(CMAKE_CXX_STANDARD 14)
+  set(CMAKE_CXX_STANDARD 17)
```
Comment (jacobperron) [inline rosidl_typesupport_c/CMakeLists.txt:12]: Is this necessary? If so, we should update the comment above.

File: rosidl_typesupport_cpp/resource/srv__type_support.cpp.em [hunk 17]
```diff
@@ -109,10 +120,70 @@ static const type_support_map_t _@(service.namespaced_type.name)_service_typesup
...
+  event_msg->info.stamp.set__nanosec(info->stamp_nanosec);
+
+  std::array<uint8_t, 16> client_id;
+  std::move(std::begin(info->client_id), std::end(info->client_id), client_id.begin());
+  event_msg->info.client_id.set__uuid(client_id);
+
+  // TODO(jacobperron): consider removing this argument and let users pass nullptr for both request and response messages
+  if (!enable_message_payload) {
+    return event_msg;
+  }
```
Comment (jacobperron) [inline rosidl_typesupport_cpp/resource/srv__type_support.cpp.em:157]: Looking for feedback on this TODO

File: rosidl_typesupport_tests/CMakeLists.txt [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+  rosidl_get_typesupport_target(cpp_typesupport_target ${PROJECT_NAME} "rosidl_typesupport_cpp")
+
+  ament_add_gtest(test_service_typesupport_cpp
+    test/rosidl_typesupport_cpp/test_service_typesupport.cpp
+  )
+  target_link_libraries(test_service_typesupport_cpp
+    "${cpp_typesupport_target}"
+  )
+
+  # TODO(jacobperron): Add tests for rosidl_typesupport_c
```
Comment (jacobperron) [inline rosidl_typesupport_tests/CMakeLists.txt:34]: We can add equivalent tests for C-typesupport to their own test file.

File: rosidl_typesupport_tests/test/rosidl_typesupport_cpp/test_service_typesupport.cpp [hunk 19]
```diff
@@ -0,0 +1,172 @@
...
+    EXPECT_EQ(event->request[0].int16_value, expected_request.int16_value);
+    EXPECT_EQ(event->request[0].uint16_value, expected_request.uint16_value);
+    ASSERT_EQ(event->response.size(), 1U);
+    EXPECT_EQ(event->response[0].bool_value, expected_response.bool_value);
+    EXPECT_EQ(event->response[0].string_value, expected_response.string_value);
+    ASSERT_TRUE(srv_ts->event_message_destroy_handle(event, &allocator));
+  }
+}
+
+// TODO(jacobperron): Add test for action service
```
Comment (jacobperron) [inline rosidl_typesupport_tests/test/rosidl_typesupport_cpp/test_service_typesupport.cpp:172]: May as well add tests with an action service.

**Review by jacobperron — COMMENTED** [review]
I've fixed the test and expanded on it. Also did some other minor cleanup.

I've left comments for things that still need to be addressed.

File: rosidl_typesupport_c/resource/srv__type_support.cpp.em [hunk 20]
```diff
@@ -105,10 +118,86 @@ static const type_support_map_t _@(service.namespaced_type.name)_service_typesup
   &_@(service.namespaced_type.name)_service_typesupport_data.data[0],
 };
 
+@#TODO(ihasdapie): Import Postfixes
+@{event_type = '__'.join([package_name, *interface_path.parents[0].parts, service.namespaced_type.name]) + SERVICE_EVENT_MESSAGE_SUFFIX}@
+@{request_type = '__'.join([package_name, *interface_path.parents[0].parts, service.namespaced_type.name]) + SERVICE_REQUEST_MESSAGE_SUFFIX}@
+@{response_type = '__'.join([package_name, *interface_path.parents[0].parts, service.namespaced_type.name]) + SERVICE_RESPONSE_MESSAGE_SUFFIX}@
+
+void *
+rosidl_typesupport_c_@('__'.join([package_name, *interface_path.parents[0].parts, service.namespaced_type.name]))__event_message__create(
```
Comment (ivanpauno) [inline rosidl_typesupport_c/resource/srv__type_support.cpp.em:127]: it sounds like we're rewriting the same function in many places, I think we need to move them to a common place

File: rosidl_typesupport_cpp/resource/srv__type_support.cpp.em [hunk 21]
```diff
@@ -109,10 +119,70 @@ static const type_support_map_t _@(service.namespaced_type.name)_service_typesup
   &_@(service.namespaced_type.name)_service_typesupport_data.data[0],
 };
 
+@{
+event_type = '::'.join([package_name, *interface_path.parents[0].parts, service.namespaced_type.name]) + SERVICE_EVENT_MESSAGE_SUFFIX
+}@
+void *
+rosidl_@('_'.join([package_name, *interface_path.parents[0].parts, service.namespaced_type.name]))_event_message_create(
```
Comment (ivanpauno) [inline rosidl_typesupport_cpp/resource/srv__type_support.cpp.em:126]: same

**Review by ivanpauno — COMMENTED** [review]
note to self: check open discussions again, review tests.

Otherwise LGTM

File: rosidl_typesupport_c/resource/srv__type_support.cpp.em (hunk 15 above)
Comment (ivanpauno) [reply rosidl_typesupport_c/resource/srv__type_support.cpp.em]: Done in https://github.com/ros2/rosidl_typesupport/pull/127/commits/f1bca7202581f53bfbacfb675b1c41d7ded85a47

File: rosidl_typesupport_cpp/resource/srv__type_support.cpp.em (hunk 17 above)
Comment (ivanpauno) [reply rosidl_typesupport_cpp/resource/srv__type_support.cpp.em:157]: I think removing the `bool` makes sense, I'm going to implement that change

File: rosidl_typesupport_c/resource/srv__type_support.cpp.em (hunk 20 above)
Comment (ivanpauno) [reply rosidl_typesupport_c/resource/srv__type_support.cpp.em:126]: Actually this is not that easy, as we need to create one function for each type.
Maybe we could generate the function once in `rosidl_generator_c/rosidl_generator_cpp` and use the correct one in each typesupport.

@wjwwood any opinion?

Comment (wjwwood) [reply rosidl_typesupport_c/resource/srv__type_support.cpp.em:126]: I'm afraid I don't understand what you mean (I haven't read through these yet). We need a version of this for each service type and language (C/C++), which part are you hoping to de-duplicate?

Comment (ivanpauno) [reply rosidl_typesupport_c/resource/srv__type_support.cpp.em:126]: We're generating one per each typesupport.
One here, same thing for typesupport fastrtps (c and cpp), and same thing for typesupport introspection (c and cpp).

It should be one per language, not one per typesupport library.
I think that having the common implementation generated in `rosidl_generator_c`/`rosidl_generator_cpp` makes sense.
Then the typesupport implementation would only need to add that to the typesupport struct.

Comment (ivanpauno) [reply rosidl_typesupport_c/resource/srv__type_support.cpp.em]: See https://github.com/ros2/rosidl/pull/700/commits/0d6f816ce0a3f86566757294a419a2dcd7965ba4 (from https://github.com/ros2/rosidl/pull/700) and https://github.com/ros2/rosidl_typesupport/pull/127/commits/61d32cb9f1cfa5e8599017e468aee58bb7b6016c.
`rosidl_generator_c` was already generating the typesupport headers, so if there's a common function for C typesupports  it makes sense to generate it there.

I'm planning to do the same for the CPP typesupport soon.

File: rosidl_typesupport_c/CMakeLists.txt (hunk 16 above)
Comment (ivanpauno) [reply rosidl_typesupport_c/CMakeLists.txt]: Undone in https://github.com/ros2/rosidl_typesupport/pull/127/commits/fb0258fcd7466695c2c50db4f87492198a36d220

**Review by wjwwood — COMMENTED** [review]
lgtm, with test todo's addressed

**Comment by ivanpauno** [comment]
> lgtm, with test todo's addressed

Tests completed in https://github.com/ros2/rosidl_typesupport/pull/127/commits/f27383168ce98e490813e59ce3bdd82323349b55.

**Review by mjcarroll — APPROVED** [review]

**Comment by clalancette** [comment]
CI for this is in https://github.com/ros2/ros2/issues/1285#issuecomment-1404070715
